mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,828 @@
+
+    /// Generate a new MetaDataHeader based on the partition information and the
+    /// underlying block_size
+    pub fn new(block_size: u32, partition: &GptEntry) -> MetaDataHeader {
+        let index_start = Aligned::get_blocks(
+            MetaDataHeader::METADATA_HEADER_SIZE,
+            block_size,
+        );
+
+        let data_start = index_start
+            + Aligned::get_blocks(
+                MetaDataHeader::MAX_INDEX_ENTRIES
+                    * MetaDataHeader::INDEX_ENTRY_SIZE,
+                block_size,
+            );
+
+        MetaDataHeader {
+            signature: [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61],
+            header_size: MetaDataHeader::METADATA_HEADER_SIZE,
+            self_checksum: 0,
+            generation: 0,
+            self_lba: partition.ent_start + 1, /* skip the first block of the
+                                                * partition */
+            index_start: index_start as u64,
+            used_entries: 0,
+            max_entries: MetaDataHeader::MAX_INDEX_ENTRIES,
+            entry_size: MetaDataHeader::INDEX_ENTRY_SIZE,
+            index_checksum: 0,
+            data_start: data_start as u64,
+            data_end: partition.ent_end - partition.ent_start - 1,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Clone)]
+pub struct MetaDataIndexEntry {
+    /// Current object revision
+    pub revision: u64,
+    /// When this entry was created (number of microseconds since UNIX_EPOCH)
+    pub timestamp: u128,
+    /// CRC-32 checksum of all the blocks comprising the stored object
+    pub data_checksum: u32,
+    /// Location of first data block (relative to self_lba in the
+    /// MetaDataHeader)
+    pub data_start: u64,
+    /// Location of last data block (relative to self_lba in the
+    /// MetaDataHeader)
+    pub data_end: u64,
+}
+
+impl MetaDataIndexEntry {
+    /// Convert a slice into an index array
+    pub fn from_slice(
+        slice: &[u8],
+        used_entries: u32,
+    ) -> Result<Vec<MetaDataIndexEntry>, MetaDataError> {
+        let mut index: Vec<MetaDataIndexEntry> = Vec::new();
+        let mut reader = Cursor::new(slice);
+        for _ in 0 .. used_entries {
+            index.push(
+                deserialize_from(&mut reader).context(DeserializeError {})?,
+            );
+        }
+        Ok(index)
+    }
+
+    /// Calculate checksum of an index array
+    pub fn checksum(index: &[MetaDataIndexEntry]) -> u32 {
+        let mut digest = crc32::Digest::new(crc32::IEEE);
+        for entry in index {
+            digest.write(&serialize(entry).unwrap());
+        }
+        digest.sum32()
+    }
+}
+
+#[derive(Debug, PartialEq, Default, Clone)]
+pub struct NexusMetaData {
+    pub header: MetaDataHeader,
+    pub index: Vec<MetaDataIndexEntry>,
+}
+
+impl NexusMetaData {
+    /// Construct a MetaDataHeader from raw data
+    fn read_header(buf: &DmaBuf) -> Result<MetaDataHeader, MetaDataError> {
```

Comment:
nit; when I read the function name I assume it "reads" from disk, but its already read from disk as its int he DmaBuf...